pages/react-canvas-components.mdx

```diff
@@ -0,0 +1,256 @@
+
+- Or, we can create a `useCanvas` hook to hide this complexity away! Even when consuming both React contexts, the hook can only return the canvas rendering context to the child components:
+
+```jsx
+export const useCanvas = () => {
+  React.useContext(FrameContext);
+  const renderingContext = React.useContext(CanvasContext);
+  return renderingContext;
+};
+```
+
+The `Hexagon` component logic now needs a small update to use this new hook:
+
+```jsx
+const Hexagon = props => {
+  const renderingContext = useCanvas();
+  // drawing logic
+};
+```
+
+### 🔄 Making the hexagons move
+
+If we want the hexagons to rotate, each hexagon should have its rotation angle change at every frame-by incrementing it by 1 for example-. To do this, the hexagons needs to _remember_ the rotation from the previous render. We will use a [React ref](/react-refs-evolution) to achieve this:
+
+```jsx
+const animatedRotation = React.useRef(props.rotation);
+animatedRotation.current = animatedRotation.current + 1;
+```
+
+As the `Hexagon` component re-renders at every frame, this code makes its rotation  change at every frame: they rotate! 😍
+
+Like we did with `useCanvas`, we can also improve a bit the readability of this code by hiding the implementation detail – here, using a React ref – in a hook:
+
+```jsx animated-hexagons-v9j2q src/Canvas.js
+const useAnimation = (initialValue, valueUpdater) => {
+  const animatedValue = React.useRef(initialValue);
+  animatedValue.current = valueUpdater(animatedValue.current);
+  return animatedValue.current;
+};
+```
+
+The `Hexagon` code now looks a bit better:
+
+```jsx animated-hexagons-v9j2q src/Hexagon.js
+const Hexagon = props => {
+  // [...]
+  const animatedRotation = useAnimation(props.rotation, angle => angle + 1);
+  // drawing logic
+};
+```
+
+👏 Congratulations 👏 We now have animated canvas-based React components! We even created two custom hooks along the way to make our code nicer.
+
+## 👀 Going further
+
+- At some point it can become quite CPU heavy to try to animate a lot of shapes in a canvas. At this point, there's another more performant way to draw things: WebGL. It's a huge subject on its own! I would recommend you to look at how to write WebGL based components with [react-three-fiber](https://github.com/react-spring/react-three-fiber) _— spoiler alert, they made their own React reconciler_
+
+- Our `useAnimation` hook is not helping if we want to animate stuff other than infinitely changing numbers. To create more complex ones, [react-spring](https://github.com/react-spring/react-spring) is the library to go. It needs a bit of wiring to have it work with our own frame loop – [here's how it's done to work with `react-three-fiber` for example](https://github.com/react-spring/react-spring/blob/bdf8afb8b2d837fd756f8ffaae6931fb15c0e3d2/src/targets/three/index.ts#L24-L25).
```

Comment:
- "Our `useAnimation` hook does not help if ... "
- "To create more complex things" 
- "library to go to" 
- "to make it work" 
- "Here's how to get it working with ... "